### DISCUSSION TITLE
next/image calculated sizes attribute

### DISCUSSION THREAD

**Discussion by strebl** [vercel/next.js#37690] [Help]
Hi, I'd like to get some input from the next.js / react community about the way we are using responsive images in our current projects. As a small background, we are currently using Vue.js / NuxtJS and looking into switching to next.js with our agency.
For that reason, we started a project using next.js and are getting familiar with it.

One part is the image handling. We are delighted with next/image because it's a well-designed, polished and feature rich component. But there is something we implemented in our custom vue component, we can't see anyone doing in the next.js world.

We calculate the sizes attribute dynamically based on the image and viewport width. We start with `sizes="1px"` and calculate the actual size on load:
```js
this.sizes = Math.ceil(image.getBoundingClientRect().width / window.innerWidth*100)+'vw'
```
We do the same on window resize. Additionally, our smallest image in the srcset is an inlined blurred svg which represents the actual image.

With that solution, we can get superfast images which start blurred, and then the closest image gets loaded without needing to specify the `sizes` attribute.

next/image has that blurred image feature, which works great. But we don't see an obvious, clean way to calculate and set the sizes attribute dynamically because we don't have access to the image element. However, the component seems so well put together and thought out and got us thinking we are doing something "wrong" 😄 
That said, we are using this solution for multiple years now and didn't encounter any problems so far.

I'm very excited to hear feedback from this community. Please tell me if I need to explain our solution better.

Just for reference, Imgix has something similar:
https://blog.imgix.com/2021/07/21/auto-sizes

**Comment by 7iomka** [comment]
@strebl Hi
Have you found a way to automate the `sizes` calculation?
For example I have this
```css
.imageContainer {
  position: relative;
  width: 88px;
  height: 94px;

  @include-media (">=xl") {
    width: 98px;
    height: 105px;
  }
}
```

```tsx
 <div className={styles.imageContainer}>
      <Image
        src={images.top ?? '/static/images/no-image.jpg'}
        alt=""
        fill
        className="object-cover rounded-sm"
        sizes="100vw"
        quality={100}
      />
</div>

```

I set sizes to `100vw` by default, but this not optimize this small image at all, and if the original image is big (1920*1080) - next.js will load the fullsize image.
It would be very nice to understand what calculations I need to make in such a case

**Comment by 7iomka** [comment]
Manual calculation looks like this
```tsx
            <div className="relative w-88 h-94 xl:w-98 xl:h-105">
              <Image
                src={images.top ?? '/static/images/no-image.jpg'}
                alt=""
                fill
                className="object-cover rounded-sm"
                sizes={`${largerThan('xl')} ${rem(98)}, ${rem(98)}`}
                quality={100}
              />
```
This is the example for container with static width values.
With percentage width values, especially if this container is also located in the area with not full width, it is imposible to make common calculations based on `vw` I think.
